dali/operators/decoder/audio/audio_decoder_op.cc

```diff
@@ -99,95 +93,67 @@ AudioDecoderCpu::SetupImpl(std::vector<OutputDesc> &output_desc, const workspace
-    size_t ret = decoders_[sample_idx]->Decode(as_raw_span(tmp_buf.data(), num_samples));
-    DALI_ENFORCE(ret == num_samples,
-                 make_string("Error decoding audio file: ", files_names_[sample_idx]));
-
-    if (should_downmix) {
-      if (should_resample) {
-        // downmix and resample
-        float *downmixed = tmp_buf.data() + meta.length * meta.channels;
-        assert(downmixed + meta.length <= tmp_buf.data() + tmp_buf.size());
-        kernels::signal::Downmix(downmixed, tmp_buf.data(), meta.length, meta.channels);
-        resampler_.Resample(audio.data, 0, audio.shape[0], output_rate,
-                            downmixed, meta.length, meta.sample_rate);
-      } else {
-        // downmix only
-        kernels::signal::Downmix(audio.data, tmp_buf.data(), meta.length, meta.channels);
-      }
-    } else if (should_resample) {
-      // multi-channel resample
-      resampler_.Resample(audio.data, 0, audio.shape[0], output_rate,
-                          tmp_buf.data(), meta.length, meta.sample_rate, meta.channels);
-
-    } else {
-      // convert or copy only - this will only happen if resampling is specified, but this
-      // recording's sampling rate and number of channels coincides with the target
-      int64_t len = std::min<int64_t>(volume(audio.shape), meta.length*meta.channels);
-      for (int64_t ofs = 0; ofs < len; ofs++) {
-        audio.data[ofs] = ConvertSatNorm<OutputType>(tmp_buf[ofs]);
-      }
-    }
-  } else {
-    assert(!should_downmix && !should_resample);
-    size_t num_samples = volume(audio.shape);
-    size_t ret = decoders_[sample_idx]->Decode(as_raw_span(audio.data, num_samples));
-    DALI_ENFORCE(ret == num_samples,
-                 make_string("Error decoding audio file: ", files_names_[sample_idx]));
-  }
+  int64_t decode_scratch_sz = 0;
+  int64_t resample_scratch_sz = 0;
+  if (should_resample || should_downmix || !std::is_same<OutputType, DecoderOutputType>::value)
+    decode_scratch_sz = meta.length * meta.channels;
+
+  // resample scratch is used to prepare a single or multiple (depending if
+  // downmixing is needed) channel float input, required by the resampling
+  // kernel
+  int64_t out_channels = should_downmix ? 1 : meta.channels;
+  if (should_resample)
+    resample_scratch_sz = meta.length * out_channels;
+
+  auto &scratch_decoder = scratch_decoder_[thread_idx];
+  scratch_decoder.Resize({decode_scratch_sz});
+
+  auto &scratch_resampler = scratch_resampler_[thread_idx];
+  scratch_resampler.Resize({resample_scratch_sz});
+
+  DecodeAudio<OutputType, DecoderOutputType>(
+    audio, *decoders_[sample_idx], meta, resampler_,
+    {scratch_decoder.mutable_data<DecoderOutputType>(), decode_scratch_sz},
+    {scratch_resampler.mutable_data<float>(), resample_scratch_sz},
+    target_sr, downmix_,
+    files_names_[sample_idx].c_str());
 }
 
-template <typename OutputType>
+template <typename OutputType, typename DecoderOutputType>
```

Comment:
Why is there a need to have separate `DecoderOutputType`? Aren't we decoding always to `float`?

Reply:
No, if we want the output in lets say int16 and you don't require resampling or downmixing, then it's better to decode to int16 directly

Reply:
It's what's always been there.